design-proposals/crd-based-management.md

```diff
@@ -0,0 +1,417 @@
+
+### User Story: Access to Cloud Resource
+
+As an application owner, I would like to extend use of an **external** (i.e., non-k8s) Cloud Service
+ to other clusters where my application runs. I would like to ensure the access is protected and
+ limited only to workloads I manage in remote locations.
+
+### User Story: Service Load Balancing and Failover
+
+As an application owner, I would like to allow a Service I use to be provided by multiple remote
+ locations. Selecting the remote endpoint should allow load balancing among the locations, with a
+ specific cluster acting as the primary location but allowing failover to other locations when
+ the primary is unavailable.
+
+## Proposal and Design Details
+
+### Namespaces and RBAC
+
+ClusterLink shall be deployed into a designated namespace (e.g., `clusterlink-system` ),
+ separate from user accessible namespaces, where workloads run. The use of a designated
+ namespace allows the definition of roles and permissions needed to manage the system.
+ The ClusterLink namespace holds objects used by the site administrator, including `Peer`
+ and `AdminPolicy`, and has the relevant RBAC configuration needed. These object will
+ only be watched for in the designated namespace and ignored in other namespaces (note:
+ to allow future extension to multiple fabrics, the control plane shall watch for these
+ objects in its local namespace and not a hard-coded namespace).
+
+Application owners can define `Export`, `Import` and `Policy` objects in namespaces
+ where their workloads run. In addition, application owners require read access to
+ `Peer` objects in the ClusterLink namespace, as `Import`s need to reference a remote
+ cluster by its name.
+
+### Custom Resource Definitions
+
+All ClusterLink CRD objects are derived from the current API structures defined, with
+ changes and adaptations needed by CRD schemas. All objects are namespace scoped.
+ Where possible, CRDs shall follow k8s recommendations for defining and managing
+ objects (e.g., clear separation of Spec and Status fields, where Spec is set by the user
+ and Status by ClusterLink controllers, use of sub-resources, etc).
+ Spec parts declare user intent and the system's actual state is reconciled with
+ the desired state asynchronously, over time. Thus, users should not expect immediate
+ feedback and state change. Other relevant practices should be followed as much as possible
+ (e.g., addition of Finalizers where appropriate, reconciliation retrieving up to date
+ information instead of using internal caches for actual state, etc,).
+
+Objects should use Conditions for tracking their Status. The following conventions for
+ status conditions are suggested. Controllers should use the following conditions types
+ to signal whether a resource has been reconciled, or if it encountered any problems:
+
+- `Reconciling`: Indicates that the resource actual state does not yet match its desired
```

Comment:
I am not sure if reconciling is the best word in deployment they use progressing